components/camel-dynamic-router/src/main/java/org/apache/camel/component/dynamicrouter/DynamicRouterMulticastProcessor.java

```diff
@@ -0,0 +1,292 @@
+     */
+    private final Supplier<PrioritizedFilter.PrioritizedFilterFactory> filterProcessorFactorySupplier;
+
+    /**
+     * Flag to log a warning if a message is dropped due to no matching filters.
+     */
+    private final boolean warnDroppedMessage;
+
+    public DynamicRouterMulticastProcessor(String id, CamelContext camelContext, Route route, String recipientMode,
+                                           final boolean warnDroppedMessage,
+                                           final Supplier<PrioritizedFilter.PrioritizedFilterFactory> filterProcessorFactorySupplier,
+                                           ProducerCache producerCache, AggregationStrategy aggregationStrategy,
+                                           boolean parallelProcessing, ExecutorService executorService,
+                                           boolean shutdownExecutorService,
+                                           boolean streaming, boolean stopOnException,
+                                           long timeout, Processor onPrepare, boolean shareUnitOfWork,
+                                           boolean parallelAggregate) {
+        super(camelContext, route, new ArrayList<>(), aggregationStrategy, parallelProcessing, executorService,
+              shutdownExecutorService,
+              streaming, stopOnException, timeout, onPrepare,
+              shareUnitOfWork, parallelAggregate);
+        setId(id);
+        this.producerCache = producerCache;
+        this.filterMap = new TreeMap<>();
+        this.recipientMode = recipientMode;
+        this.filterProcessorFactorySupplier = filterProcessorFactorySupplier;
+        this.warnDroppedMessage = warnDroppedMessage;
+    }
+
+    public boolean isIgnoreInvalidEndpoints() {
+        return ignoreInvalidEndpoints;
+    }
+
+    public void setIgnoreInvalidEndpoints(boolean ignoreInvalidEndpoints) {
+        this.ignoreInvalidEndpoints = ignoreInvalidEndpoints;
+    }
+
+    protected List<Processor> createEndpointProcessors(Exchange exchange) {
+        List<String> recipientList = matchFilters(exchange).stream()
+                .map(PrioritizedFilter::getEndpoint)
+                .distinct()
+                .map(String::trim)
+                .collect(Collectors.toList());
+        if (recipientList.isEmpty()) {
+            // No matching filters, so we will use the default filter that will create a
+            // notification that there were no routing participants that matched the
+            // exchange, which results in a "dropped" message.
+            Message exchangeIn = exchange.getIn();
+            Object originalBody = exchangeIn.getBody();
+            exchangeIn.setHeader("originalBody", originalBody);
+            String endpoint = String.format(LOG_ENDPOINT, this.getClass().getCanonicalName(), getId(),
+                    warnDroppedMessage ? "WARN" : "DEBUG");
+            recipientList.add(endpoint);
+            String error = String.format(
```

Comment:
Hmmm why is this a plain string message returned and not an exception or something, its a bit weird

Reply:
The idea is that, at any given time, recipients could be there, or they could be absent.  If one participant sends a message and there are no recipients for a message, then that message is logged.  In configuration, it can be specified to set this log level to `WARN`.  Otherwise, it is just logged at the debug level.  In a subsequent ticket, I could add the option to set an exception on the exchange when there are no valid recipients present for a message.